src/goLanguageServer.ts

```diff
@@ -88,18 +94,38 @@ export async function registerLanguageFeatures(ctx: vscode.ExtensionContext) {
 			},
 			revealOutputChannelOn: RevealOutputChannelOn.Never,
 			middleware: {
-				provideDocumentFormattingEdits: (document: vscode.TextDocument, options: FormattingOptions, token: vscode.CancellationToken, next: ProvideDocumentFormattingEditsSignature) => {
-					if (!config.features.format) {
-						return [];
+				didSave: (data: vscode.TextDocument, next: NextSignature<vscode.TextDocument, void>) => {
+					// If the user only wants diagnostics on save, show the diagnostics now.
+					if (config.features.diagnostics == DiagnosticFrequency.OnSave) {
+						const diagnostics = latestDiagnostics.get(data.uri.toString());
```

Comment:
Why not use the `uri` as key for the map and avoid the string conversions when setting and getting the values?